protelis/protelis-interpreter/src/main/java/org/protelis/vm/impl/SimpleExecutionEnvironment.java

```diff
@@ -14,6 +17,7 @@
 public final class SimpleExecutionEnvironment implements ExecutionEnvironment {
 
     private final Map<FasterString, Object> env = new LinkedHashMap<>();
+    private final Set<String> keys = new HashSet<>();
```

Comment:
Is there any reason for maintaining a second set? Since it is backed by an `HashMap`, its maintenance cost is computationally the same of switching `env` key type to `String`. I would do so, as I see no valuable performance advantage (the costly part is hashing the string the first time if there are many different strings), and as such keeping two data structures sounds very much like a premature optimization/code smell.

Moreover in case we don't want to change the implementation, let me recommend `LinkedHashSet` (everywhere), as `HashSet` has no guarantee on reproducibility of iteration order, and that is source of unpredictability when running simulations (breaks reproducibility).

Just switching `env` to `private final Map<String, Object> env = new LinkedHashMap<>();` would solve also this issue I believe.